common/memstream.h

```diff
@@ -342,6 +342,96 @@ class MemoryReadWriteStream : public SeekableReadStream, public SeekableWriteStr
 	byte *getData() { return _data; }
 };
 
+/**
+ * A seekable read and writeable memory stream that operates on an already existing memory buffer
+ */
+class MemorySeekableReadWriteStream : public SeekableReadStream, public SeekableWriteStream {
+private:
+	const uint32 _bufSize;
+	byte *_ptrOrig;
+	byte *_ptr;
+	uint32 _pos;
+	bool _err;
+	bool _eos;
+
+public:
+	MemorySeekableReadWriteStream(byte *buf, uint32 len) : _ptrOrig(buf), _ptr(buf), _bufSize(len), _pos(0), _err(false), _eos(false) {}
+
+	uint32 write(const void *dataPtr, uint32 dataSize) override {
+		// Write at most as many bytes as are still available...
+		if (dataSize > _bufSize - _pos) {
+			dataSize = _bufSize - _pos;
+			// We couldn't write all the data => set error indicator
+			_err = true;
+		}
+		memcpy(_ptr, dataPtr, dataSize);
+		_ptr += dataSize;
+		_pos += dataSize;
+		return dataSize;
+	}
+
+	int32 pos() const override { return _pos; }
+	int32 size() const override { return _bufSize; }
+
+	bool eos() const override { return _eos; }
+
+	bool err() const override { return _err; }
+	void clearErr() override { _err = false; }
+
+	inline void rewind(int32 bytes) {
+		seek(pos() - bytes);
+	}
+
+	byte peekByte() {
+		if (_bufSize - _pos <= 0) {
+			_eos = true;
+			return 0;
+		}
+		return *_ptr;
+	}
+
+	uint32 read(void *dataPtr, uint32 dataSize) override {
+		// Read at most as many bytes as are still available...
+		if (dataSize > _bufSize - _pos) {
+			dataSize = _bufSize - _pos;
+			_eos = true;
+		}
+		memcpy(dataPtr, _ptr, dataSize);
+
+		_ptr += dataSize;
+		_pos += dataSize;
+
+		return dataSize;
+	}
+
+	bool seek(int32 offset, int whence = SEEK_SET) override {
+		switch (whence) {
+		case SEEK_END:
+			// SEEK_END works just like SEEK_SET, only 'reversed',
+			// i.e. from the end.
+			offset = size() + offset;
+			// Fall through
+		case SEEK_SET:
+			// Fall through
+		default:
+			_ptr = _ptrOrig + offset;
+			_pos = offset;
+			break;
+		case SEEK_CUR:
+			_ptr += offset;
+			_pos += offset;
+			break;
+		}
+		// Post-Condition
+		if ((int32)_pos > size()) {
+			_pos = size();
+			_ptr = _ptrOrig + _pos;
+		}
+
+		return true;
```

Comment:
Since a successful `seek` is supposed to reset the end-of-stream flag, I think you need to set `_eos = false;` before returning here.

Reply:
Nice catch - thanks a lot. Fixed now.